nixos/modules/services/web-apps/bookstack.nix

```diff
@@ -215,238 +366,120 @@ in {
-      APP_URL = cfg.appURL;
-      DB_HOST = db.host;
-      DB_PORT = db.port;
-      DB_DATABASE = db.name;
-      DB_USERNAME = db.user;
-      MAIL_DRIVER = mail.driver;
-      MAIL_FROM_NAME = mail.fromName;
-      MAIL_FROM = mail.from;
-      MAIL_HOST = mail.host;
-      MAIL_PORT = mail.port;
-      MAIL_USERNAME = mail.user;
-      MAIL_ENCRYPTION = mail.encryption;
-      DB_PASSWORD._secret = db.passwordFile;
-      MAIL_PASSWORD._secret = mail.passwordFile;
-      APP_SERVICES_CACHE = "/run/bookstack/cache/services.php";
-      APP_PACKAGES_CACHE = "/run/bookstack/cache/packages.php";
-      APP_CONFIG_CACHE = "/run/bookstack/cache/config.php";
-      APP_ROUTES_CACHE = "/run/bookstack/cache/routes-v7.php";
-      APP_EVENTS_CACHE = "/run/bookstack/cache/events.php";
-      SESSION_SECURE_COOKIE = tlsEnabled;
-    };
-
-    environment.systemPackages = [ artisan ];
-
-    services.mysql = mkIf db.createLocally {
-      enable = true;
-      package = mkDefault pkgs.mariadb;
-      ensureDatabases = [ db.name ];
-      ensureUsers = [
-        { name = db.user;
-          ensurePermissions = { "${db.name}.*" = "ALL PRIVILEGES"; };
-        }
-      ];
-    };
+  config = lib.mkIf cfg.enable {
 
     services.phpfpm.pools.bookstack = {
-      inherit user;
-      inherit group;
+      inherit user group;
+      phpPackage = cfg.package.phpPackage;
       phpOptions = ''
         log_errors = on
         post_max_size = ${cfg.maxUploadSize}
         upload_max_filesize = ${cfg.maxUploadSize}
       '';
       settings = {
-        "listen.mode" = "0660";
-        "listen.owner" = user;
-        "listen.group" = group;
+        "listen.mode" = lib.mkDefault "0660";
+        "listen.owner" = lib.mkDefault user;
+        "listen.group" = lib.mkDefault group;
+        "pm" = lib.mkDefault "dynamic";
+        "pm.max_children" = lib.mkDefault 32;
+        "pm.start_servers" = lib.mkDefault 2;
+        "pm.min_spare_servers" = lib.mkDefault 2;
+        "pm.max_spare_servers" = lib.mkDefault 4;
+        "pm.max_requests" = lib.mkDefault 500;
       } // cfg.poolConfig;
     };
 
-    services.nginx = {
-      enable = mkDefault true;
+    services.nginx = lib.mkIf (cfg.nginx != null) {
+      enable = true;
       recommendedTlsSettings = true;
       recommendedOptimisation = true;
       recommendedGzipSettings = true;
-      virtualHosts.${cfg.hostname} = mkMerge [ cfg.nginx {
-        root = mkForce "${bookstack}/public";
-        locations = {
-          "/" = {
-            index = "index.php";
-            tryFiles = "$uri $uri/ /index.php?$query_string";
-          };
-          "~ \\.php$".extraConfig = ''
-            fastcgi_pass unix:${config.services.phpfpm.pools."bookstack".socket};
-          '';
-          "~ \\.(js|css|gif|png|ico|jpg|jpeg)$" = {
-            extraConfig = "expires 365d;";
+      virtualHosts.${cfg.hostname} = lib.mkMerge [
+        cfg.nginx
+        {
+          root = lib.mkForce "${cfg.package}/public";
```

Comment:
Why use mkForce here?

Reply:
Where is the other root defined?

Reply:
Not in this module. It is defined in the [vhost-options](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-servers/nginx/vhost-options.nix) submodule.